pkg/remediator/watch/watcher.go

```diff
@@ -17,60 +17,48 @@ package watch
 import (
 	"context"
 
-	"k8s.io/apimachinery/pkg/api/meta"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/watch"
-	"k8s.io/client-go/dynamic"
 	"k8s.io/client-go/rest"
 	"kpt.dev/configsync/pkg/declared"
 	"kpt.dev/configsync/pkg/remediator/queue"
 	"kpt.dev/configsync/pkg/status"
 )
 
-type startWatchFunc func(metav1.ListOptions) (watch.Interface, error)
-
 // watcherConfig contains the options needed
 // to create a watcher.
 type watcherConfig struct {
 	gvk                     schema.GroupVersionKind
-	mapper                  meta.RESTMapper
 	config                  *rest.Config
 	resources               *declared.Resources
 	queue                   *queue.ObjectQueue
 	scope                   declared.Scope
 	syncName                string
-	startWatch              startWatchFunc
+	startWatch              WatchFunc
 	addConflictErrorFunc    func(status.ManagementConflictError)
 	removeConflictErrorFunc func(status.ManagementConflictError)
 }
 
-// createWatcherFunc is the type of functions to create watchers
-type createWatcherFunc func(ctx context.Context, cfg watcherConfig) (Runnable, status.Error)
-
-// createWatcher creates a watcher for a given GVK
-func createWatcher(ctx context.Context, cfg watcherConfig) (Runnable, status.Error) {
-	if cfg.startWatch == nil {
-		mapping, err := cfg.mapper.RESTMapping(cfg.gvk.GroupKind(), cfg.gvk.Version)
-		if err != nil {
-			return nil, status.APIServerErrorf(err, "watcher failed to get REST mapping for %s", cfg.gvk.String())
-		}
+// watcherFactory knows how to build watch.Runnables.
+type watcherFactory func(ctx context.Context, cfg watcherConfig) (Runnable, status.Error)
 
-		dynamicClient, err := dynamic.NewForConfig(cfg.config)
-		if err != nil {
-			return nil, status.APIServerErrorf(err, "watcher failed to get dynamic client for %s", cfg.gvk.String())
-		}
-
-		if cfg.scope == declared.RootReconciler {
-			cfg.startWatch = func(options metav1.ListOptions) (watch.Interface, error) {
-				return dynamicClient.Resource(mapping.Resource).Watch(ctx, options)
-			}
-		} else {
-			cfg.startWatch = func(options metav1.ListOptions) (watch.Interface, error) {
-				return dynamicClient.Resource(mapping.Resource).Namespace(string(cfg.scope)).Watch(ctx, options)
+// watcherFromListerWatcher returns a new watcherFactory that uses the specified
+// ListerWatcherFactory.
+func watcherFromListerWatcher(factory ListerWatcherFactory) watcherFactory {
+	factoryPtr := factory
+	return func(ctx context.Context, cfg watcherConfig) (Runnable, status.Error) {
```

Comment:
It doesn't look like the context input of this function is used. The context from the nested `startWatch` function seems to be used instead. Should the `watcherFactory` signature be changed to not include a context input arg?

Reply:
good catch. removed.